src/tools/idlc/src/idl.c

```diff
@@ -18,145 +18,210 @@
-   %nounistd only disables inclusion in the generated source file. */
-#define YY_NO_UNISTD_H
-#include "idl.l.h"
+#include "parser.h"
+#include "typetree.h"
 #include "tt_create.h"
 
 #include "dds/ddsrt/heap.h"
 #include "dds/ddsrt/log.h"
 #include "dds/ddsrt/misc.h"
 #include "dds/ddsrt/string.h"
-#include "dds/ddsts/typetree.h"
 
-idl_parser_t *idl_create_parser(void)
+int32_t idl_processor_init(idl_processor_t *proc)
 {
-  idl_parser_t *parser;
-
-  if ((parser = ddsrt_calloc(1, sizeof(*parser))) != NULL &&
-      (parser->context = ddsts_create_context()) != NULL &&
-      (idl_yylex_init(&parser->yylstate)) == 0 &&
-      (parser->yypstate = (void *)idl_yypstate_new()) != NULL)
-  {
-    return parser;
+  memset(proc, 0, sizeof(*proc));
+  if (!(proc->context = ddsts_create_context())) {
+    return IDL_MEMORY_EXHAUSTED;
+  } else if (!(proc->parser.yypstate = (void *)idl_yypstate_new())) {
+    ddsts_free_context(proc->context);
+    return IDL_MEMORY_EXHAUSTED;
   }
 
-  idl_destroy_parser(parser);
-  return NULL;
+  return 0;
 }
 
-void idl_destroy_parser(idl_parser_t *parser)
+void idl_processor_fini(idl_processor_t *proc)
 {
-  if (parser != NULL) {
+  if (proc) {
     idl_file_t *file, *next;
 
-    if (parser->yypstate != NULL) {
-      idl_yypstate_delete((idl_yypstate *)parser->yypstate);
-    }
-    if (parser->yylstate != NULL) {
-      idl_yylex_destroy((yyscan_t)parser->yylstate);
-    }
-    if (parser->context != NULL) {
-      ddsts_free_context(parser->context);
-    }
+    if (proc->parser.yypstate)
+      idl_yypstate_delete((idl_yypstate *)proc->parser.yypstate);
+    if (proc->context)
+      ddsts_free_context(proc->context);
 
-    for (file = parser->files; file != NULL; file = next) {
+    for (file = proc->files; file; file = next) {
       next = file->next;
-      if (file->name != NULL) {
+      if (file->name)
         ddsrt_free(file->name);
-      }
       ddsrt_free(file);
     }
 
-    if (parser->buffer.data != NULL) {
-      ddsrt_free(parser->buffer.data);
-    }
-
-    ddsrt_free(parser);
+    if (proc->buffer.data)
+      ddsrt_free(proc->buffer.data);
   }
 }
 
-dds_return_t idl_scan_token(idl_parser_t *parser)
+static void
+idl_log(
+  idl_processor_t *proc, uint32_t prio, idl_location_t *loc, const char *fmt, va_list ap)
+{
+  char buf[1024];
+  size_t off;
+  ssize_t cnt;
+
+  (void)proc;
+  (void)prio;
+  if (loc->first.file)
+    cnt = snprintf(
+      buf, sizeof(buf), "%s:%u:%u: ", loc->first.file, loc->first.line, loc->first.column);
+  else
+    cnt = snprintf(
+      buf, sizeof(buf), "%u:%u: ", loc->first.line, loc->first.column);
+
+  if (cnt == -1)
+    return;
+
+  off = (size_t)cnt;
+  cnt = vsnprintf(buf+off, sizeof(buf)-off, fmt, ap);
+
+  if (cnt == -1)
+    return;
+
+  off += (size_t)cnt;
+
+  if (off >= sizeof(buf))
+    buf[sizeof(buf) - 1] = '\0';
+  else
+    buf[off] = '\0';
+
+  fprintf(stderr, "%s\n", buf);
```

Comment:
Consider `fputs`

Reply:
Leaving it as is for now as `fputs` doesn't append a `\n`.